Feature/new anonymous circuit

lucasmenendez opened the pull request:
As Vocdoni, we want to be able to have fully anonymous voting available in the protocol/SDK/UIs to allow the voters to have full privacy. 

**Protocol/Backend**: Implement `Secret Identity Key (SIK)` in the Vochain
- [x] Create the new transactions for SIK: Update and create needed protobuffer definitions (more info [here](https://hackmd.io/8QtTh6FGTn6FlYgiQ1Ubxg?view#Tech-specification)) --> PR [here](https://github.com/vocdoni/dvote-protobuf/pull/63).
- [x] Keeping track of the starting blocks of each open election storing into a no-state database the `startBlock` of every started election. If a election is finished, delete it from the database. If a election is paused, the database will store the current block (height) (more info [here](https://hackmd.io/8QtTh6FGTn6FlYgiQ1Ubxg#SIK-reset)).
- [x] Create new Merkle Tree in the Vochain state ([here](https://github.com/vocdoni/vocdoni-node/blob/main/vochain/state/trees.go#L120)) with similar logic that the `RollingCensus` but using `Poseidon` hash ([current example](https://github.com/vocdoni/vocdoni-node/blob/main/vochain/state/rollingcensus.go))
- [x] Store `SIK` roots for hysteresis into statedb (nostate)
- [x] Implement the anonymous voting transaction flow with root hysteresis.
  - [x] Support to store multiple `electionID`s by `blockNumber` on ProcessBlockRegistry.
    - [x] Unit testing for ProcessBlockRegistry.
  - [x] Check validity of sikRoot.
- [x] API endpoints and APIClient methods to manage `SIK`'s:
  - [x] Update account metadata setter to handle `SIK` during creation and update the account ([API](https://github.com/vocdoni/vocdoni-node/blob/ed06c3517c0560312eac3ed105f827b00b1efc9f/api/accounts.go#L139) & [APIClient](https://github.com/vocdoni/vocdoni-node/blob/ed06c3517c0560312eac3ed105f827b00b1efc9f/apiclient/account.go#L155))
  - [x] Create API endpoint and APIClient method to delete account `SIK`.
  - [x] Update API endpoints and APIClient methods to retrieve not just the census proof (root + siblings) for an election, but also the `SIK` proof (root + siblings).
- [x] Review leaf insertion in both trees (`SIK` and census) to ensure that follows something similar to this [example](https://github.com/vocdoni/zk-franchise-proof-circuit/blob/9a2c24e5c0fdddc77f65cac16a9e411dabeb1725/internal/helpers.go#L39).
- [x] Update nullifier formula and implementation to include user `SIK` 
- [x] Deprecate and remove the ZkAddress.
- [x] Update circuit reference and artifacts URL's.
- [x] Adapt end2end tests.
- [x] Expose the list of valid SIK roots on a new API endpoint.
- [x] Split the sikproof from the censusproof API endpoint and move everything related to the SIK to a new path (`/sik`) (read more [here](https://hackmd.io/_tGmlKE-SluYsuP10Ts33g)).
- [x] Truncate the sik signature to 64 bytes discarding the byte number 65 (for recovery) that produces differences between the signatures generated with go and javascript.
- [x] Create a free tx to allow to non registered users that are in the census of a created election to register a sik without create a vochain account. This tx will have no cost.

**Important TODO's before merge**
- [x] Set new tx's costs on staging
- [x] Change resource of the API endpoints related to the SIK to plural (`/sik` => `/siks`).
- [x] Change the hysteresis end2end test to unit test to allow to set the chain height manually and optimize the test
- [x] Update the API endpoint to get account info to include if the account has a valid SIK registered.
- [x] Keep track of the number of SIKs registered with RegisterSIKTx by process to prevent reaching the MaxCensusSize number. Remove the record when the election ends.

lucasmenendez committed:
initial branch with SIK subtree configuration

lucasmenendez committed:
initial sik subtree state methods

lucasmenendez committed:
require the hysteresis height as parameter on DelSIK method

lucasmenendez committed:
hysteresis value encoding decoding methods and tests

lucasmenendez committed:
initializing sik tree on state, new errors and SetSIK method and test, fixing SIK encoding

lucasmenendez committed:
new errors and DelSIK method

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,128 @@
+
+// encodedHeightLen constant is the number of bytes of the encoded hysteresis
+// that contains the hysteresis height value, starting from the last byte
+const encodedHeightLen = 4
```
Join both consts in a single statement

lucasmenendez replied:
Both are used independently

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,128 @@
+//   - If the hysteresis height is not reached yet, it returns an error.
+//   - If the hysteresis height is reached, it updates the value of the sik with
+//     the provided one.
+func (v *State) SetSIK(address common.Address, newSik []byte) error {
```
I think SetSik should just set the sik. All the logic remaining hysteresis and the conditions to reset an existing sik should be part of the APP logic, not the state. Specifically part of the transaction handler.

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,128 @@
+	return v.Tx.DeepSet(address.Bytes(), encodeHysteresis(hysteresisHeight), StateTreeCfg(TreeSIK))
+}
+
+// encodeHysteresis funtion returns the encoded value of the height hysteresis
```
This approach is not correct. It does not avoid double voting.

Instead of an hysteresis height, we should store the the current maximum endBlock for a process.

lucasmenendez committed:
fixing DelSIK method and test

altergui commented on vochain/state/errors.go:
```diff
@@ -12,4 +12,8 @@ var (
+	ErrRegisteredValidSIK   = fmt.Errorf("address already has a valid sik")
+	ErrSIKAlreadyInvalid    = fmt.Errorf("sik is already invalidated")
+	ErrHysteresisNotReached = fmt.Errorf("hysteresis height not reached")
+	ErrInvalidHysteresis    = fmt.Errorf("hysteresis must be greater than the curren height")
```
s/curren/current/

altergui commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,122 @@
+// that contains the hysteresis height value, starting from the last byte
+const encodedHeightLen = 4
+
+// hysteresisLen constant contains the number of bytes that the encode
```
maybe `s/encode/encoded/` ?

altergui commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,122 @@
+		v.Tx.Lock()
+		defer v.Tx.Unlock()
+		return v.Tx.DeepAdd(address.Bytes(), newSik, StateTreeCfg(TreeSIK))
+	} else if err != nil {
```
no need for "else", you can simply do two separate ifs

altergui commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,122 @@
+	for len(bHeight) < encodedHeightLen {
+		bHeight = append([]byte{0}, bHeight...)
+	}
+	// create the hysteresis with the rigth number of zeros
```
right

altergui commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,122 @@
+
+// decodeHeigh funtion returns the decoded height uint32 from the leaf
+// value that contains the encoded hysteresis height.
+func decodeHeigh(leafValue []byte) uint32 {
```
s/decodeHeigh/decodeHeight/

altergui replied:
(also in func doc comment, and in callers)

altergui reviewed (commented):
reviewed but i'm not familiar with the concepts/idea so i'll leave the approval to p4u, commenting here just some typos

lucasmenendez committed:
methods to keep track of the startblocks of opened processes

lucasmenendez committed:
fixing tests and some small changes and typos

lucasmenendez committed:
improving sik tests

lucasmenendez committed:
new UpdateSIKRoots method to keep the track of the last n sik merkle tree roots to support voting of already registered voters during a election where new users are registered

lucasmenendez committed:
update siks roots on sik creation, update or deletion, sik roots testing

lucasmenendez committed:
some no-state logic deleted from sik, reimplementation using custom type and updated tests

altergui commented on vochain/state/process.go:
```diff
@@ -161,6 +164,52 @@ func (v *State) CountProcesses(committed bool) (uint64, error) {
+	return processesTree.NoState().Delete(pid)
+}
+
+// MinOnGoingStartBlock returns the minimun start block of the current on going
```
`ongoing` is a word, not two.
so this would be MinOngoingStartBlock

also, `minimum` is the correct spelling

lucasmenendez replied:
Sorry about my english ;(

altergui commented on vochain/state/process.go:
```diff
@@ -211,9 +260,28 @@ func (v *State) UpdateProcess(p *models.Process, pid []byte) error {
+	if err := updateProcess(&v.Tx, p, pid); err != nil {
+		return err
+	}
+	v.Tx.Unlock()
```
if `err != nil` happens, Unlock will never happen. add it before the `return err` , or use defer, if the `try to update startBlocks database` doesn't take long

altergui replied:
or even simpler
```go
 	v.Tx.Lock()
	err := updateProcess(&v.Tx, p, pid)
    v.Tx.Unlock()
	if err != nil {
		return err
	}
```

lucasmenendez replied:
this code is old, but very good point, it will be fixed in the next commit

lucasmenendez committed:
fixing some comments of new methods regarding processes start blocks

lucasmenendez commented on vochain/state/process.go:
```diff
@@ -211,9 +264,28 @@ func (v *State) UpdateProcess(p *models.Process, pid []byte) error {
+		return err
+	}
+	v.Tx.Unlock()
+	// try to update startBlocks database
```
@p4u I dont know if this file is the best place for this code or if it needs to be moved to `transaction/transaction.go` because its not about the state at all.

p4u replied:
LGTM

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
support for available weight and vote weight vote data and model attributes

lucasmenendez committed:
initial transaction logic to support set and delete SIK's

altergui commented on api/census_test.go:
```diff
@@ -225,9 +225,10 @@ func TestCensusProof(t *testing.T) {
-					LeafWeight: censusData.Value,
+					Type:            models.ProofArbo_BLAKE2B,
+					Siblings:        censusData.Proof,
+					AvalaibleWeight: censusData.Value,
```
typo here: should be `Available`

lucasmenendez committed:
fixing typo

lucasmenendez committed:
include function to get the maximun endBlock from current processes to avoid to invalidate a SIK before the correct height to avoid double voting

lucasmenendez committed:
split set sik and del sik txs

lucasmenendez committed:
fixing recursion bug calling the same method from itself

lucasmenendez committed:
include the option to set a SIK during account creation with apiclient

lucasmenendez committed:
new API type, endpoints and APIClient methods to update and delete SIK

lucasmenendez committed:
wrong place for invalidated sik height checks (wrong delSik, correct setSik)

lucasmenendez committed:
renaming proof to censusProof and introducing sikProof mechanisms on api

lucasmenendez committed:
including sikRoot on censusProof api endpoint

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
modify the method to get sik proof to get encoded siblings, include sik siblings as part of census proof response

lucasmenendez committed:
create new function for the new nullifier definition

p4u commented on api/accounts.go:
```diff
@@ -407,3 +425,30 @@ func (a *API) tokenTransfersHandler(msg *apirest.APIdata, ctx *httprouter.HTTPCo
+//	@Description	Updates the Secret Identity Key of an Address
+//	@Success		200	{object}	SikSet
+//	@Router			/accounts/sik [post]
+func (a *API) setSikHandler(msg *apirest.APIdata, ctx *httprouter.HTTPContext) error {
```
I'm not sure of the need of this API method. It does nothing different from `/chain/transaction`.

We do have specific endpoints for some transactions that require special logic. Such as new election, which requires the Gateway to publish the metadata into IPFS. But for the rest of transactions, we use the generic endpoint which just relays the transaction.

lucasmenendez replied:
Perfect, it makes sense

p4u commented on api/api_types.go:
```diff
@@ -228,18 +228,26 @@ type AccountSet struct {
 	MetadataURL string         `json:"metadataURL" swaggertype:"string"`
 }
 
+type SikSet struct {
```
No need either.

lucasmenendez committed:
removing no needed api endpoints and replace with generic chain transactions requests, set protobuf repository origin to the correct dev branch, including new methods on state treeviews and updaters to handel circom siblings and removing no needed logic from vochain app

lucasmenendez committed:
new function to represent number in a field, fixing almost all tests except anonymous that got some issues with the lenght of trees keys

p4u commented on apiclient/account.go:
```diff
@@ -107,6 +107,16 @@ func (c *HTTPclient) AccountBootstrap(faucetPkg *models.FaucetPackage, metadata
 	}
 
+	if sik == nil {
+		sikSign, err := c.account.SignEthereum([]byte(DefaultSIKContent))
```
I think the crypto/ethereum package could be extended to have a method that returns and abstracts the SIK generation. So we can do `c.account.Sik()`.

p4u commented on apiclient/account.go:
```diff
@@ -227,3 +238,95 @@ func (c *HTTPclient) GetTransfers(from common.Address, page, pageSize int) ([]*i
+
+// SetSik function allows to update the Secret Identity Key for the current
+// HTTPClient account. To do that, the function requires a secret user input.
+func (c *HTTPclient) SetSik(secret []byte) (types.HexBytes, error) {
```
secret is not used :/

I'd export two methods, this might be useful for testing purposes:

- SetSik()
- SetCustomSik(sik []byte)

The first can call the second.

p4u commented on apiclient/helpers.go:
```diff
@@ -314,5 +320,49 @@ func UnmarshalFaucetPackage(data []byte) (*models.FaucetPackage, error) {
+//	SIK = poseidon(address, signature, secret*)
+//
+// *The secret is optional.
+func (c *HTTPclient) GenerateSik(sign, secret []byte) ([]byte, error) {
```
I don't think we need this HTTPclient method. 

The SIK generation should be only part of the crypto/ethereum package. Let's abstract it.

p4u commented on vochain/state/process.go:
```diff
@@ -161,6 +162,92 @@ func (v *State) CountProcesses(committed bool) (uint64, error) {
 	return count, nil
 }
 
+// RegisterStartBlock function creates a new record on the key-value database
```
This is a method, not a function. 

Please, specify this is being stored in the no-state database

p4u commented on vochain/state/process.go:
```diff
@@ -161,6 +162,92 @@ func (v *State) CountProcesses(committed bool) (uint64, error) {
+// RegisterStartBlock function creates a new record on the key-value database
+// associated to the processes subtree for the process ID and the startBlock
+// number provided.
+func (v *State) RegisterStartBlock(pid []byte, startBlock uint32) error {
```
I think the name of the method is not self-descriptive. I'd be more explicit: `ProcessRegisterStartBlockNoState()`

p4u replied:
Actually I think we might make a better structure to clearly differentiate this methods, since its hard to understand what is the actual purpose.

To this end I'd propose to create a new type: `ProcessBlockRegistry`

On the `State` constructor, we can create a new instance for this type:

`State.ProcessBlockRegistry = ProcessBlockRegistry{ treeRW: processTree.NoState(), treeRead: ....,  treeLock: v.Tx }`

Then whenever you want to use it, just:

`state.ProcessBlockRegistry.SetStartBlock(pid, startBlock)`

So all methods are contained inside the ProcessBlockRegistry.

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,247 @@
+
+// SIKByAddress function return the current SIK value associated to the provided
+// address.
+func (v *State) SIKByAddress(address common.Address) (SIK, error) {
```
SIKfromAddress()

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,247 @@
+//     value provided.
+//   - If it exists but it is not valid, overwrite the stored value with the
+//     provided one.
+func (v *State) SetSIK(address common.Address, newSik SIK) error {
```
SetAddressSik()

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,247 @@
+
+// DecodeInvalidatedHeight funtion returns the decoded height uint32 from the
+// leaf value that contains an invalidated SIK.
+func (s SIK) DecodeInvalidatedHeight() uint32 {
```
Move this methods next to the type definition

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,247 @@
+//   - If it does not exist, remove all roots with a lower block number except
+//     for the next lower sikRoot. It is becouse it still being validate for a
+//     period.
+func (v *State) UpdateSIKRoots() error {
```
I wonder if for SIKs we can follow the same approach as in ProcessBlockRegistry. 
My feeling is that the State package is growing too much and some hierarchy would be appreciated.

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,247 @@
+	sikRootsDB := siksTree.NoState()
+	// calculate the current minimun block to purge useless sik roots
+	currentBlock := v.CurrentHeight()
+	if currentBlock > SIKROOT_HYSTERESIS_BLOCKS {
```
`if currentBlock := v.CurrentHeight(); currentBlock > SIKROOT_HySTERESIS_BLOCS {`

lucasmenendez replied:
it can't be, it is used later

p4u commented on vochain/transaction/proof.go:
```diff
@@ -115,21 +115,21 @@ func VerifyProofOffChainTree(process *models.Process, proof *models.Proof,
 
-		factoryWeight := arbo.BytesToBigInt(p.LeafWeight)
-		if p.VotingWeight == nil {
+		factoryWeight := arbo.BytesToBigInt(p.AvailableWeight)
```
Maybe it is a good time to change factoryWeight by the new naming

lucasmenendez replied:
done, i will check the rest of the code to ensure that we use the same naming

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +239,42 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+		return fmt.Errorf("invalid transaction")
+	}
+
+	bAddress := tx.GetAddress()
```
This is wrong, you cannot trust the Address available in the Protobuf model is correct.

Actually the address should not be a field, instead you should extract the address from the Signature.

Please check initializeSignedVote() function. You need to do something similar.

p4u replied:
Let's ensure no other TxCheck method works like this... either we have a huge security hole.

p4u commented on vochain/transaction/transaction.go:
```diff
@@ -434,6 +439,41 @@ func (t *TransactionHandler) CheckTx(vtx *vochaintx.Tx, forCommit bool) (*Transa
 		}
 
+	case *models.Tx_SetSik:
+		err := t.SetSikTxCheck(vtx)
```
this method might return the address and the newSik, so you don't need to recalculate/refetch it again.

altergui commented on Dockerfile:
```diff
@@ -7,6 +7,7 @@ ARG BUILDARGS
 # Build all the binaries at once, so that the final targets don't require having
 # Go installed to build each of them.
 WORKDIR /src
+# COPY . .
```
?

do note that instead of doing the COPY, a bind mount is used in line 17
`--mount=type=bind,target=.`

lucasmenendez replied:
Do you know if this works with docker-compose too? or only with docker?

altergui replied:
both. docker-compose uses docker to build images

altergui commented on api/api_types.go:
```diff
@@ -222,25 +222,27 @@ type Account struct {
+	TxPayload   []byte         `json:"txPayload,omitempty" swaggerignore:"true"`
+	Metadata    []byte         `json:"metadata,omitempty" swaggerignore:"true"`
+	TxHash      types.HexBytes `json:"txHash" `
+	MetadataURL string         `json:"metadataURL" swaggertype:"string"`
```
why do you need `swaggertype:"string"` if the type is already a string?

altergui commented on apiclient/helpers.go:
```diff
@@ -343,3 +345,24 @@ func (c *HTTPclient) GenerateSik(sign, secret []byte) ([]byte, error) {
+//
+//	nullifier = poseidon(signature, secret*, sha256(electionId))
+//
+// *The secret is optional.
```
nit: the `*` has a special meaning in golang so i'd avoid using it for other meanings. you can simply mention "secret", it will be evident you're referring to the parameter, no need for `*`

what does "optional" mean? parameters are never optional like in other languages

my suggestion

```go
//  nullifier = poseidon(signature, secret, sha256(electionId))
//
// secret can be nil, in that case blah will happen
```

altergui commented on crypto/zk/utils.go:
```diff
@@ -20,6 +20,19 @@ const (
 	publicSigLen = 7
 )
 
+var modulus, _ = new(big.Int).SetString("21888242871839275222246405745257275088548364400416034343698204186575808495617", 10)
```
seeing this number made me 21888242871839275222246405745257275088548364400416034343698204186575808495617 times curious to know where does it come from. please add a comment?

lucasmenendez replied:
hahaha, for sure, thanks a lot!

altergui reviewed (commented):
i'm still completely lost on the actual code, so i'm just contributing some trivial comments based on "blind" proofreading

lucasmenendez committed:
* ZkAddress deprecated and deleted, now the voter key is the ethereum address
* New zk circuit and input generation updated (wip)
* Sik and Nullifier getters moved to crypto/ethereum
* Removed GetCircomSiblings method from tree, censustree and statedb
* ProofToCircomSiblings function created in crypto/zk package
* Process block state methods moved to new ProcessBlockRegistry struct
* Tests and benchmarks updated
* Bug fixes and renaming

lucasmenendez committed:
* Merge branch 'main' into feature/new_anonymous_circuit
* State.Sik method splitted in both Sik() and CustomSik(secret) methods
* Small changes

lucasmenendez committed:
* more detailed errors in zk prover
* sik generation fixed
* circom siblings encoding fixed
* circuit inputs generation fixed
* tests and benchmarks updated
* anonymous e2e test fails (wip)

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
mod tidy

lucasmenendez committed:
fixing end2end test

altergui commented on crypto/zk/prover/prover.go:
```diff
@@ -105,7 +105,7 @@ func (p *Proof) Nullifier() (*big.Int, error) {
 	}
 	// Get the nullifier from the fourth public signal of the proof
-	strNullifier := p.PubSignals[3]
+	strNullifier := p.PubSignals[2]
```
fourth? or third?

lucasmenendez committed:
including sikroot verification during anonymous voting, new e2e test to check hysteresis, small bugs fixed

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
overwrite dvote-proto dep to the branch which contains last changes

lucasmenendez committed:
fixing tests and proof inputs decoding

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
go mod tidy

lucasmenendez committed:
unit tests for ProcessRegistryBlock, new API endpoint to return the currently valid SIK roots

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
new circuit input naming

lucasmenendez committed:
transaction_zk_test improved, inputs and proof generated during the test instead to be hardcoded

mariajdab commented on cmd/end2endtest/hysteresis.go:
```diff
@@ -0,0 +1,129 @@
+		return err
+	}
+
+	log.Info("wating to reach hysteresis to create new account and force to delete old sikroots")
```
typo `wating`

mariajdab commented on cmd/end2endtest/hysteresis.go:
```diff
@@ -0,0 +1,129 @@
+		return err
+	}
+	testAccountPrivKey = testAccount.PrivateKey()
+	if _, _, err := t.createAccount(testAccountPrivKey.String()); err != nil {
```
I would suggest due this code is repeated twice, you could use a anonymous function: 
```
	forceUpdateSikRoots := func() error {
		log.Info("wating to reach hysteresis to create new account and force to delete old sikroots")

		ctx, cancel := context.WithTimeout(context.Background(), t.config.timeout)
		defer cancel()
		if err := t.api.WaitUntilNBlocks(ctx, state.SIKROOT_HYSTERESIS_BLOCKS); err != nil {
			return err
		}
		log.Info("create to account to force to update sik roots")
		testAccount := ethereum.NewSignKeys()
		if err := testAccount.Generate(); err != nil {
			return err
		}
		testAccountPrivKey := testAccount.PrivateKey()
		if _, _, err := t.createAccount(testAccountPrivKey.String()); err != nil {
			return err
		}
		return nil
	}
```
It's not worth putting it in the helpers since it won't be used again, I think.

mariajdab commented on vochain/state/processblockregistry.go:
```diff
@@ -0,0 +1,83 @@
+	mainTree      statedb.TreeViewer
+}
+
+// RegisterStartBlock method creates a new record on the key-value database
```
Just a quick remember to keep the comments starting whit the function o method name

mariajdab commented on vochain/state/processblockregistry.go:
```diff
@@ -0,0 +1,83 @@
+	return pbr.startBlocksDB.Delete(pid)
+}
+
+// MinReadyProcessStartBlock returns the minimun start block of the current on going
```
I think this needs to be updated too

mariajdab commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,274 @@
+		// if not exists create it
+		log.Debugw("setSIK (create)",
+			"address", address.String(),
+			"sik", SIK(newSik).String())
```
If I am not wrong `newSik` is already a SIK type (type of the argument function)

lucasmenendez committed:
census proof endpoint does not return the sik proof any more, it has been moved to the new /sik api path with the old /chain/info/sikroots (now /sik/roots), tests updated and some feedback applied

lucasmenendez committed:
SikRoot API struct replaced with Census struct

lucasmenendez committed:
hotfix: return circuit hash as string instead of as base64 uintarray

lucasmenendez committed:
new method created on crypt/ethereum to get a signature compatible with the sik generation (discarding the last byte)

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
update vochain genesis version

lucasmenendez committed:
fixing errors during sik transactions checks and delivers, now the sik is optional during create account processes

lucasmenendez committed:
import duplicated

lucasmenendez committed:
new genesis version for dev

lucasmenendez committed:
split sik transactions checkers instead of call between them, enforce to setSik and delSik txs to have a cost, new free tx for register a sik without account, proving that the address is already registered into a election

lucasmenendez committed:
update protobuf dependency to the new version and update delSik and setSik tx on the apiclient

lucasmenendez committed:
make deepsource linter happy

lucasmenendez committed:
make gofmt linter happy

lucasmenendez committed:
make deepsource linter happy: comment for the hysteresis number of blocks constant

lucasmenendez committed:
make deepsource linter happy

lucasmenendez committed:
setting sik txs costs for dev environment

lucasmenendez committed:
updating tx cost test with new tx names

p4u commented on crypto/ethereum/ethereum.go:
```diff
@@ -182,6 +188,18 @@ func (k *SignKeys) SignVocdoniMsg(message []byte) ([]byte, error) {
+// SignEthereum method, using the DefaultSikPayload and discarding the last
+// byte of the signature (used for recovery), different that the same byte of a
+// signature generated with javascript.
+func (k *SignKeys) SignVocdoniSik() ([]byte, error) {
```
I'd call it something such as `SIKsignature()`

p4u commented on crypto/ethereum/ethereum.go:
```diff
@@ -196,6 +214,66 @@ func (k *SignKeys) VerifySender(message, signature []byte) (bool, ethcommon.Addr
+//	SIK = poseidon(address, signature, secret)
+//
+// The secret could be nil.
+func (k *SignKeys) CustomSik(secret []byte) ([]byte, error) {
```
For clarity, I'd rather call it `AccountSIK(secret []byte)` and remove the Sik() method bellow. 
The caller can just specify secret=nil.

p4u commented on crypto/ethereum/ethereum.go:
```diff
@@ -196,6 +214,66 @@ func (k *SignKeys) VerifySender(message, signature []byte) (bool, ethcommon.Addr
+
+// Nullifier method composes the nullifier of the current SignKeys for the
+// desired election id and the secret provided.
+func (k *SignKeys) Nullifier(electionId, secret []byte) ([]byte, error) {
```
We should specify this is actually the SIK nullifier for the current account.  So I'd propose `AccountSIKnullifier()`

p4u commented on crypto/zk/circuit/inputs.go:
```diff
@@ -12,45 +14,71 @@ import (
-	factoryWeight, votingWeight *big.Int, censusSiblings []string) (*CircuitInputs, error) {
-	if zkAddr == nil || censusRoot == nil || electionId == nil ||
-		factoryWeight == nil || votingWeight == nil || len(censusSiblings) == 0 {
+func GenerateCircuitInput(account *ethereum.SignKeys, password, electionId,
```
Create a struct out of these parameters.

p4u commented on vochain/state/process.go:
```diff
@@ -211,9 +211,28 @@ func (v *State) UpdateProcess(p *models.Process, pid []byte) error {
+	if err := updateProcess(&v.Tx, p, pid); err != nil {
+		return err
+	}
+	v.Tx.Unlock()
```
The Unlock should be deferred as it was before, else in case of an error on updateProcess, the mutex would be locked forever.

p4u commented on vochain/state/processblockregistry.go:
```diff
@@ -0,0 +1,83 @@
+	"go.vocdoni.io/dvote/statedb"
+)
+
+type ProcessBlockRegistry struct {
```
Please document what the ProcessBlockregistry does and its purpose

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,279 @@
+
+// SIKFromAddress function return the current SIK value associated to the provided
+// address.
+func (v *State) SIKFromAddress(address common.Address) (SIK, error) {
```
SIKfromAddress according to camelCase rules

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,279 @@
+// ValidSIKRoots returns the list of current valid roots from the SIK's merkle
+// tree. It reads the roots from the key-value database associated to the SIK's
+// subtree.
+func (v *State) ValidSIKRoots() ([][]byte, error) {
```
This looks like a query that we might perform several times on the same block, in case there are several votes attempting the same election. 

Same as we do with Height, we might store the ValidSIKRoots() slice into memory on app.BeginBlock(). Then any query from the application or transactionHandler should use the cached version of `app.SIKvalidRoots()`. Specially the ones checking the validity of a vote.

Take this into consideration.

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,279 @@
+	if err != nil {
+		return err
+	}
+	return v.UpdateSIKRoots()
```
Review all namings that do not follow the camel case rules, please :)

https://capitalizemytitle.com/camel-case/

lucasmenendez replied:
According with this page and others that I've seen, the correct form is `UpdateSIKRoots` or `UpdateSikRoots` (but the last one brakes our *rule* of keep SIK in uppercase)

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,279 @@
+
+// ExpiredSik returns if the provided siksRoot is still valid or not, checking
+// if it is included into the list of current valid sik roots.
+func (v *State) ExpiredSik(candidateRoot []byte) (bool, error) {
```
We should always use capital letters for `SIK`

lucasmenendez replied:
Also on protobuf messages?

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,279 @@
+	defer v.Tx.Unlock()
+	siksTree, err := v.Tx.DeepSubTree(StateTreeCfg(TreeSIK))
+	if err != nil {
+		return err
```
Please, try to wrap the errors (done in my proposed function)

p4u commented on vochain/state/sik.go:
```diff
@@ -0,0 +1,279 @@
+//   - If it does not exist, remove all roots with a lower block number except
+//     for the next lower sikRoot. It is becouse it still being validate for a
+//     period.
+func (v *State) UpdateSIKRoots() error {
```
You can simplify this function by doing a single iteration.

```
func (v *State) UpdateSIKRoots() error {
	// get sik roots key-value database associated to the siks tree
	v.Tx.Lock()
	defer v.Tx.Unlock()

	siksTree, err := v.Tx.DeepSubTree(StateTreeCfg(TreeSIK))
	if err != nil {
		return fmt.Errorf("error getting deep sub tree: %w", err)
	}
	newSikRoot, err := siksTree.Root()
	if err != nil {
		return fmt.Errorf("error getting siksTree root: %w", err)
	}

	sikRootsDB := siksTree.NoState()

	currentBlock := v.CurrentHeight()
	if currentBlock > SIKROOT_HYSTERESIS_BLOCKS {
		minBlock := currentBlock - SIKROOT_HYSTERESIS_BLOCKS
		minBlockKey := make([]byte, 32)
		binary.LittleEndian.PutUint32(minBlockKey, minBlock)

		var toPurge [][]byte
		var nearestLowerBlock uint32

		sikRootsDB.Iterate(nil, func(key, value []byte) bool {
			candidateKey := bytes.Clone(key)
			blockNumber := binary.LittleEndian.Uint32(candidateKey)
			if blockNumber < minBlock {
				if _, err := sikRootsDB.Get(minBlockKey); err == nil || blockNumber > nearestLowerBlock {
					toPurge = append(toPurge, candidateKey)
					nearestLowerBlock = blockNumber
				}
			}
			return true
		})

		for _, blockToDelete := range toPurge {
			if err := sikRootsDB.Delete(blockToDelete); err != nil {
				return fmt.Errorf("error deleting block: %w", err)
			}
			log.Debugw("updateSIKRoots (deleted)",
				"blockNumber", binary.LittleEndian.Uint32(blockToDelete))
		}
	}

	log.Debugw("updateSIKRoots (created)",
		"newSikRoot", hex.EncodeToString(newSikRoot),
		"blockNumber", currentBlock)

	blockKey := make([]byte, 32)
	binary.LittleEndian.PutUint32(blockKey, currentBlock)

	if err := sikRootsDB.Set(blockKey, newSikRoot); err != nil {
		return fmt.Errorf("error setting new sikRoot: %w", err)
	}

	return nil
}
```

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -15,95 +15,96 @@ import (
 )
 
 // CreateAccountTxCheck checks if an account creation tx is valid
```
I'm uncertain if it is worth it to return the SIK, the function does not perform any operation on it, so there won't be extra cost if tx.GetSIK() is executed by the caller (instead of the function).

lucasmenendez replied:
In fact, `CheckTx` function does not need the SIK unless the Tx is for commit, so we can call `tx.GetSIK()` just in this case.

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
 }
+
+// DelSikTxCheck checks if a delete sik tx is valid
+func (t *TransactionHandler) DelSikTxCheck(vtx *vochaintx.Tx) (common.Address, error) {
```
Shouldn't we check if the SIK can actually be deleted?  So it must exist already and not be invalidated.

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+	// check if the address already has invalidated sik to ensure that it is
+	// not updated after reach the correct height to avoid double voting
+	if currentSik, err := t.state.SIKFromAddress(txAddress); err == nil {
+		maxEndBlock, err := t.state.ProcessBlockRegistry.MaxEndBlock(t.state.CurrentHeight(), false)
```
Please use app.Height() instead

lucasmenendez replied:
It seems that i have not access to the app here... But probably i am wrong

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+	}
+	// check if the address already has invalidated sik to ensure that it is
+	// not updated after reach the correct height to avoid double voting
+	if currentSik, err := t.state.SIKFromAddress(txAddress); err == nil {
```
You should check that err != nil and different to arbo.KeyNotFound

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+	return txAddress, nil
+}
+
+// SetSikTxCheck checks if a set sik tx is valid
```
Explain how this function allows or discards a potential sik change.

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+		return common.Address{}, nil, fmt.Errorf("cannot get the account for the tx address: %w", err)
+	}
+	newSik := tx.GetSik()
+	if newSik == nil {
```
Do the "cheap" checks before the expensive.

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+	}
+	// check if the address already has invalidated sik to ensure that it is
+	// not updated after reach the correct height to avoid double voting
+	if currentSik, err := t.state.SIKFromAddress(txAddress); err == nil {
```
I'd not allow to change the SIK on this function, to simplify and reduce potential attack vectors. I'd just allow register if the SIK does not exist at all (not even invalidated).

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+	if _, err := t.state.SIKFromAddress(txAddress); err == nil {
+		return txAddress, nil, fmt.Errorf("this address already has a valid sik")
+	}
+	// get census proof and election information provided
```
I'd do all the gets and basic checks (such as the electionID field exist) at the beginning of this function. Since all of these are cheap checks.

p4u commented on vochain/transaction/account_tx.go:
```diff
@@ -239,3 +240,133 @@ func (t *TransactionHandler) SetAccountInfoTxCheck(vtx *vochaintx.Tx) error {
+	if electionId == nil {
+		return common.Address{}, nil, fmt.Errorf("no proof provided")
+	}
+	election, err := t.state.Process(electionId, false)
```
Warning! You should check the election is in READY state

p4u commented on vochain/transaction/transaction.go:
```diff
@@ -195,22 +195,22 @@ func (t *TransactionHandler) CheckTx(vtx *vochaintx.Tx, forCommit bool) (*Transa
 
```
Pau: keep review from here

lucasmenendez committed:
fixing error with anonymous election results

lucasmenendez committed:
* Some state SIK methods renamed
* UpdateSIKRoots reimplemented
* New errors on state SIK
* Small fixes

lucasmenendez committed:
removing demo file

lucasmenendez committed:
fixing error where statedb keeps locked for ever, make secret optional during SIK generation, new comments on anonymous flow during vote_tx

p4u commented on vochain/transaction/vote_tx.go:
```diff
@@ -129,40 +129,46 @@ func (t *TransactionHandler) VoteTxCheck(vtx *vochaintx.Tx, forCommit bool) (*vs
+		// check if it is expired
+		if expired, err := t.state.ExpiredSIK(proofSIKRoot); err != nil {
+			return nil, fmt.Errorf("error checking sik root expiration")
+		} else if expired {
```
you can avoid the else

lucasmenendez committed:
envolve parameters to generate circuit inputs into a struct, comments for ProcessBlockRegistry

lucasmenendez committed:
format code, modify CreateAccountTxCheck to not return the SIK, SIK API paths changed to plural

lucasmenendez committed:
chahange all 'Sik' or 'sik' occurrences by 'SIK', fixing some bugs on checkers of SIK Tx's

lucasmenendez committed:
valid SIK roots cached at the beginning of each block to reduce state queries

lucasmenendez committed:
fixing #1058 panic

lucasmenendez committed:
ProcessBlockRegistry records prefixed on NoState DB allowing to use it to other records safely

lucasmenendez committed:
SIK records prefixed on NoState DB allowing to use it to other records safely

lucasmenendez committed:
fixing error introduced prefixing the wrong db key on SIKFromAddress state method, now the /accounts/{address} API endpoint includes the SIK

jordipainan committed:
vochaintx: fix nil pointer on unmarshal proto

lucasmenendez committed:
updating circuit inputs and proof encoding and decoding to get new voteWeight public signal

lucasmenendez committed:
Now the state keeps track of the number of SIKs registered with RegisterSIKTx by process to prevent reaching the MaxCensusSize number and prevent attacks. This counter is removed when the election ends.

lucasmenendez committed:
new apiclient function to register SIK for free if the account is already registered in an election census

lucasmenendez committed:
new test for hysteresis, end2end test removed, new costs for the new txs on stage

lucasmenendez committed:
new functions to persist the relation between a SIK and an election

lucasmenendez committed:
new models of protobuf with tempSIKs attribute on Processes message, relation between SIK's and processes created on RegisterSIKTx if the process has tempSIKs setted to true

lucasmenendez committed:
introducing some checkers to arbo.UnpackSiblings trying to prevent out of bounds panics, more info on #1058

lucasmenendez committed:
Merge branch 'main' into feature/new_anonymous_circuit

lucasmenendez committed:
new API endpoint to check if an address has a valid SIK already registered

lucasmenendez committed:
returning the sik as response of the new API endpoint

lucasmenendez committed:
force to reset env updating the genesis

altergui commented on vochain/state/sik.go:
```diff
@@ -323,6 +323,63 @@ func (v *State) CountRegisterSIK(pid []byte) (uint32, error) {
 	return binary.LittleEndian.Uint32(rawCount), nil
 }
 
+// SIKAssingTo function persists the relation between a created SIK (without
```
func name is outdated

altergui commented on tree/arbo/tree.go:
```diff
@@ -1008,21 +1008,32 @@ func PackSiblings(hashFunc HashFunction, siblings [][]byte) ([]byte, error) {
+	// to prevent runtime slice out of bounds error check if the lenght of the
+	// rest of the slice is at least equal to the encoded full length value
+	if len(b) < 4 {
 		return nil, fmt.Errorf("no enought size of packed siblings provided")
```
the `t` moved a loooong way: should be "not enough size"

but maybe it could be even `no packed siblings provided`?

altergui commented on tree/arbo/tree.go:
```diff
@@ -1008,21 +1008,32 @@ func PackSiblings(hashFunc HashFunction, siblings [][]byte) ([]byte, error) {
 // UnpackSiblings unpacks the siblings from a byte array.
 func UnpackSiblings(hashFunc HashFunction, b []byte) ([][]byte, error) {
-	if len(b) < 2 {
+	// to prevent runtime slice out of bounds error check if the lenght of the
```
`length` is misspelled in all of the comments

p4u committed:
Merge remote-tracking branch 'origin/main' into feature/new_anonymous_circuit

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
use last commit from dvote protobuf

p4u committed:
run go mod tidy

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
remove remaining rollingcensus state related functions

deepsource-autofix[bot] committed:
refactor: autofix issues in 1 file

Resolved issues in vochain/state/processblockregistry.go with DeepSource Autofix

p4u committed:
remove unused setProcessResults transaction from tx costs

p4u committed:
Merge remote-tracking branch 'origin/main' into feature/new_anonymous_circuit

altergui commented on vochain/state/processblockregistry.go:
```diff
@@ -33,7 +33,7 @@ func (pbr *ProcessBlockRegistry) DeleteStartBlock(pid []byte) error {
 // MinStartBlock returns the minimun start block of the current on going
 // processes from the no-state db associated to the process sub tree.
-func (pbr *ProcessBlockRegistry) MinStartBlock(fromBlock uint32, committed bool) (uint32, error) {
+func (pbr *ProcessBlockRegistry) MinStartBlock(fromBlock uint32, _ bool) (uint32, error) {
```
given there's no interface involved, this unused parameter sounds like a bug or an incomplete feature. same with MaxEndBlock

p4u committed:
test fix

p4u committed:
make Gui happy :)

p4u committed:
make Gui happy 2
